add rules and alertmanager sidecar to cortex-helm-chart

amouhadi opened the pull request:
This PR uses the same pattern that the official Grafana chart uses to dynamically configure dashboards and datasources and notifier. 
I tested adding both rules and alertmanager sidecar on my fork and it successfully discovered the rules and alertmanager config and got added to ruler and alertmanager.
The ConfigMaps can be put in the specified namespace and they are automatically detected and added as files to Ruler and/or AlertManager containers. This will allow easy and extensible configuration that avoids having to store state in the Cortex system itself.
This feature is disabled by default and below is an example of how it can be enabled in values.yaml
```yaml
sidecar:
  rules:
    enabled: true
    searchNamespace: cortex-rules
  alertmanager:
    enabled: true
    searchNamespace: cortex-alertmanager
```

ShuzZzle commented on docs/index.yaml:
```diff
@@ -55,8 +55,8 @@ entries:
-    - https://cortexproject.github.io/cortex-helm-chart/cortex-0.4.1.tgz
-    version: 0.4.1
+    - https://cortexproject.github.io/cortex-helm-chart/cortex-0.5.0.tgz
+    version: 0.5.0
```
Please dont bump versions on your own.

ShuzZzle replied:
Also remove the 0.5.0.tgz please

amouhadi replied:
I updated the CHANGELOG.md, deleted the 0.5.0.tgz and removed the version. Thanks!

ShuzZzle reviewed (changes requested):
Missing CHANGELOG.md documentation and please dont bump the version :)

ShuzZzle commented:
Hi @amouhadi 

We just merged #144. Due to the massive refactor you will have to rebase master and make sure that you don't actually add back the old files, since the location changed to a more easily maintainable folder structure

So for you should be templates/alertmanager/* and templates/ruler/*

Heads-up: We have deleted the role.yaml and the binding since we dropped support for PSP's and didn't really do anything else

amouhadi commented:
> Hi @amouhadi
> 
> We just merged #144. Due to the massive refactor you will have to rebase master and make sure that you don't actually add back the old files, since the location changed to a more easily maintainable folder structure
> 
> So for you should be templates/alertmanager/* and templates/ruler/*
> 
> Heads-up: We have deleted the role.yaml and the binding since we dropped support for PSP's and the fact that it didn't really do anything else

Thank you @ShuzZzle, I merged from master and resolved the conflicts.

amouhadi commented:
> I want to test a few more things, but looks good

hi @ShuzZzle, did you have a chance to test things?

ShuzZzle commented:
> > I want to test a few more things, but looks good
> 
> hi @ShuzZzle, did you have a chance to test things?

I am super unfimiliar with the concept and I didn't really had time to review it yet. Also it seems to me that some fields in the values.yml (such as "SCProvider") are never actually used?

ShuzZzle commented on values.yaml:
```diff
@@ -1477,3 +1477,75 @@ configsdb_postgresql:
       key:
+
+## Sidecars that collect the configmaps with specified label and stores the included files them into the respective folders
+sidecar:
```
Are the options here all actually somehow used? Doesn't seem like thats the case too me

amouhadi replied:
no the SCProvider is not used. I followed the same pattern that grafana chart is using. I removed it
Grafana is doing something with an extra config map: https://github.com/grafana/helm-charts/blob/4a7818cbff95b0308f5a8fac079f8adf5e3ec263/charts/grafana/templates/configmap-dashboard-provider.yaml
Defined as volume here: https://github.com/grafana/helm-charts/blob/f0e236156f935ccc46c70ca02ec0cb41151b0e7b/charts/grafana/templates/_pod.tpl#L461
Mounted here: https://github.com/grafana/helm-charts/blob/f0e236156f935ccc46c70ca02ec0cb41151b0e7b/charts/grafana/templates/_pod.tpl#L278

ShuzZzle commented on templates/ruler/ruler-dep.yaml:
```diff
@@ -72,6 +119,10 @@ spec:
               {{ toYaml .Values.ruler.extraVolumeMounts | nindent 12}}
             {{- end }}
+            {{- if .Values.sidecar.rules.enabled }}
+            - name: sc-rules-volume
```
This is actually dependent on having a volume "sc-rules-volume" in the volumes: section which currently doesn't exists, except if you provide it via ruler.extraVolumes

ShuzZzle replied:
Can this be of type emptyDir?

gburton1 replied:
Yes, this was missed when creating this PR--it is present on the forked branch that we are running against. We'll get the missing stuff sorted out.

ShuzZzle commented:
Another thing which is completely missing is the role-binding.

ShuzZzle reviewed (changes requested):
Can you please recheck the chart? Make sure its actually running properly without adding any custom/extra stuff in your cluster. Maybe share a very minimalistic cortex configuration so people can very easily test it out.

Also the role binding is completely missing

amouhadi commented:
> @gburton1 @amouhadi
> Let me know when the missing role-binding is in :)

Added rolebinding.yaml

ShuzZzle commented on README.md:
```diff
@@ -70,6 +70,61 @@ To upgrade Cortex use the following command:
+sidecar:
+  rules:
+    enabled: true
+    searchNamespace: cortex-rules
```
This won't really work since the role is bounded to a namespace

gburton1 replied:
I see your point here--a ClusterRole would seem to be the more appropriate choice. However, it's not clear why it's working fine for us as it is now--the ruler pod is querying that namespace without issue and mounting everything that it finds there. We'll study this one a bit more.

ShuzZzle replied:
This is 100% still a blocker for me. I have no Idea why that is working for you guys, that should **never** work unless your rules are in the namespace where you deployed your apps. See [Roles](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#role-and-clusterrole)

ShuzZzle replied:
Can you check how the other repo's does it?

gburton1 replied:
Grafana chart has a similar setup:
```
# If specified, the sidecar will search for dashboard config-maps inside this namespace.    
# Otherwise the namespace in which the sidecar is running will be used.    
# It's also possible to specify ALL to search in all namespaces    
searchNamespace: null
```
https://github.com/grafana/helm-charts/blob/main/charts/grafana/values.yaml#L630..L633

And Grafana has a Role defined just like Cortex, _but it also has a ClusterRole defined_ with the permissions to list and get configmaps, and this ClusterRole is also bound to the ServiceAccount:
```
- apiGroups: [""] # "" indicates the core API group
  resources: ["configmaps", "secrets"]
  verbs: ["get", "watch", "list"]
```
https://github.com/grafana/helm-charts/blob/main/charts/grafana/templates/clusterrole.yaml

So our likely answer here will be to convert our Role to a ClusterRole, but we'll try to understand why it's working for us now even without it. And we should be able to demonstrate this when we stand up a simple Cortex system with the ci/test-values.yaml--it should fail with 403 without the ClusterRole, and then succeed when we add it.

We'll dig into this first thing this week (the role issue and the restructuring of the .Values.sidecar variables)

amouhadi replied:
Here is the minimalistic cortex configuration based on ci/test-values.yaml with the sidecars configured:
 
```
ingress:
  enabled: true
 
config:
  ingester:
    lifecycler:
      join_after: 30s
      final_sleep: 30s
      tokens_file_path: "/data/tokens"
      ring:
        replication_factor: 1
        kvstore:
          store: "memberlist"
    walconfig:
      wal_enabled: true
      recover_from_wal: true
      wal_dir: "/data/wal"
 
  # See https://github.com/cortexproject/cortex/blob/master/docs/configuration/config-file-reference.md#storage_config
  storage:
    engine: blocks
  blocks_storage:
    backend: "filesystem"
    tsdb:
      dir: "/data/tsdb"
    bucket_store:
      sync_dir: "/data/tsdb-sync"
    filesystem:
      dir: "/data/store"
 
  distributor:
    ring:
      kvstore:
        store: memberlist
 
  ruler:
    ring:
      kvstore:
        store: memberlist
 
  ruler_storage:
    backend: "filesystem"
    filesystem:
      dir: "/data/store"
 
  alertmanager_storage:
    backend: "filesystem"
    filesystem:
      dir: "/data/store"
 
  querier:
    store_gateway_addresses: "dns+cortex-store-gateway-headless:9095"
 
  memberlist:
    join_members:
    - "cortex-memberlist.cortex"
 
ingester:
  replicas: 1
  statefulSet:
    enabled: true
distributor:
  replicas: 1
nginx:
  replicas: 1
 
alertmanager:
  sidecar:
    resources:
      limits:
        cpu: 100m
        memory: 100Mi
      requests:
        cpu: 50m
        memory: 50Mi
    enabled: true
    searchNamespace: cortex-alertmanager
ruler:
  sidecar:
    resources:
      limits:
        cpu: 100m
        memory: 100Mi
      requests:
        cpu: 50m
        memory: 50Mi
    enabled: true
    searchNamespace: cortex-rules
 
```
 
We set up the Cortex system locally in an isolated, empty Kubernetes cluster by doing the following:
```
helm repo add cortex-helm https://amouhadi.github.io/cortex-helm-chart/
helm install cortex --namespace test-cortex -f ci/test-values.yaml cortex-helm/cortex --version 0.5.5-rc.2
```
 
We then deployed a config map to the cortex-rules namespace (not the same namespace that the ruler itself is in)
```
apiVersion: v1
data:
  high-cpu.yaml: |-
    groups:
      - name: high-cpu
        rules:
          - alert: HighCPUusage
            expr: avg(100 - rate(node_cpu_seconds_total{mode="idle"}[5m]) * 100) by (instance) > 95
            for: 30m
            labels:
              severity: warning
            annotations:
              description: Metrics from {{ $labels.job }} on {{ $labels.instance }} show CPU > 95% for 30m.
              title: Node {{ $labels.instance }} has high CPU usage
kind: ConfigMap
metadata:
  annotations:
    k8s-sidecar-target-directory: /tmp/rules/fake
  creationTimestamp: "2021-05-23T03:46:06Z"
  labels:
    cortex_rules: "1"
  name: rules-cortex-9f99md47tc
  namespace: cortex-rules
  resourceVersion: "25639302"
  uid: 83fb43df-b3c8-4840-ad54-b11a4a1494f0
```
 
And then we observed in the ruler sidecar logs that it discovered the config map and processed it, so somehow it has access to it.
```
[2021-06-14 20:40:41] Working on configmap cortex-rules/rules-cortex-9f99md47tc
[2021-06-14 20:40:41] Found a folder override annotation, placing the rules-cortex-9f99md47tc in: /tmp/rules/fake
[2021-06-14 20:40:41] File in configmap high-cpu.yaml ADDED
```
 
As a final experiment, we also removed the serviceAccountName/serviceAccount from the Ruler deployment completely and redeployed the Ruler pods. The sidecar was still able to access the config map in the cortex-rules namespace.

ShuzZzle replied:
Did you guys disable RBAC, by any chance? I mean thats super odd.... This is what I am getting when I use the same config

> HTTP response headers: HTTPHeaderDict({'Cache-Control': 'no-cache, private', 'Content-Type': 'application/json', 'X-Content-Type-Options': 'nosniff', 'Date': 'Wed, 16 Jun 2021 07:55:00 GMT', 'Content-Length': '314'})
HTTP response body: b'{"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"configmaps is forbidden: User \\"system:serviceaccount:cortex-test:cortex-test\\" cannot watch resource \\"configmaps\\" in API group \\"\\" in the namespace \\"cortex-rules\\"","reason":"Forbidden","details":{"kind":"configmaps"},"code":403}\n'

amouhadi replied:
I was just testing in Docker for Mac’s built-in Kubernetes cluster with default settings, which does have RBAC on (api-server setting: --authorization-mode=Node,RBAC)
Updated to use ClusterRole and ClusterRoleBinding

ShuzZzle commented on README.md:
```diff
@@ -70,6 +70,61 @@ To upgrade Cortex use the following command:
+    local:
+      path: /data
+ruler:
+  rule_path: /data/rules
```
This also doesn't work since the storage volume in the ruler dep is mounted as /data and cortex isn't creating top-level folders

ShuzZzle replied:
> level=error ts=2021-06-11T08:45:45.141885066Z caller=mapper.go:50 msg="unable to read rules directory" path=/data/rules err="open /data/rules: no such file or directory"

gburton1 replied:
This is working fine for us...take a look under the /data folder. Everything you see there is created by the ruler process itself, including the rules folder. I'm trying to think why you are getting the different result. Maybe there was a change between Cortex 1.7 and 1.9?
```
docker-desktop# kubectl exec -it cortex-ruler-bd4bf566-hs57r -c cortex -n cortex-components -- sh
/ # ls -ltr /data
total 12
drwxr-xr-x    3 root     root          4096 Jun 11 01:39 tsdb-sync
drwxr-xr-x    3 root     root          4096 Jun 11 01:39 cortex
drwxr-xr-x    3 root     root          4096 Jun 11 01:40 rules
```

gburton1 replied:
We actually rebased our live system on the latest version of this chart today, which upgraded us to Cortex 1.9, and everything is still working quite well (no error like you are getting). So I will share our complete Cortex config on this PR, and our Helm chart value overrides, and perhaps something will be obvious. Then we'll try to get stand up the Cortex system with the minimal config in ci/test-values.yaml with our sidecar (I assume you just run that in a sandbox k8s cluster on your laptop?). We really appreciate your patience and help on this!

ShuzZzle replied:
Oh my God I know why, because before the sample alert didn't had the cortex_rules label and was not picked up by the sidecar -.-. The error is fine and happens if no rules has been created

ShuzZzle commented on README.md:
```diff
@@ -70,6 +70,61 @@ To upgrade Cortex use the following command:
+metadata:
+  annotations:
+    k8s-sidecar-target-directory: /tmp/rules/fake
+apiVersion: v1
```
this example has no namespace or metadata.name
see above comments

gburton1 replied:
OK, we'll add those back into this example. We were trying to build this readme example with the minimal config to make it as simple as possible.

ShuzZzle reviewed (commented):
Can you please share a minimalistic cortex configuration with a configured sidecar? We have a very minimalistic working example in the ci folder of this repo where you can get started.

gburton1 commented:
Cortex config:
```
alertmanager:
  external_url: /api/prom/alertmanager
  enable_api: true
  data_dir: /data/
  storage:
    type : local
    local:
      path: /data
api:
  prometheus_http_prefix: /prometheus
auth_enabled: false
store_gateway:
  sharding_enabled: true
  sharding_ring:
    kvstore:
      consul:
        consistent_reads: true
        host: consul-consul-server.cortex:8500
        http_client_timeout: 20s
blocks_storage:
  azure:
    account_key: <redacted>
    account_name: <redacted>
    container_name: cortex
    endpoint_suffix: blob.core.usgovcloudapi.net
  backend: azure
  bucket_store:
    sync_dir: /data/tsdb-sync
    max_chunk_pool_bytes: 4294967296
  tsdb:
    dir: /data/tsdb
chunk_store:
  chunk_cache_config:
    memcached:
      expiration: 1h
    memcached_client:
      timeout: 1s
  max_look_back_period: 0s
distributor:
  pool:
    health_check_ingesters: true
  shard_by_all_labels: true
frontend:
  compress_responses: true
  log_queries_longer_than: 10s
ingester:
  lifecycler:
    final_sleep: 0s
    join_after: 0s
    num_tokens: 512
    ring:
      kvstore:
        consul:
          consistent_reads: true
          host: consul-consul-server.cortex:8500
          http_client_timeout: 20s
        prefix: collectors/
        store: consul
      replication_factor: 3
  max_transfer_retries: 0
ingester_client:
  grpc_client_config:
    max_recv_msg_size: 104857600
    max_send_msg_size: 104857600
limits:
  enforce_metric_name: false
  reject_old_samples: true
  reject_old_samples_max_age: 168h
memberlist:
  join_members: []
frontend_worker:
  grpc_client_config:
    max_send_msg_size: 100000000
querier:
  active_query_tracker_dir: /data/cortex/querier
  query_ingesters_within: 12h
  store_gateway_addresses: dns+cortex-store-gateway-headless.cortex-components:9095
query_range:
  align_queries_with_step: true
  cache_results: true
  results_cache:
    cache:
      memcached:
        expiration: 1h
      memcached_client:
        timeout: 6s
  split_queries_by_interval: 24h
ruler:
  enable_alertmanager_discovery: false
  enable_api: true
  rule_path: /data/rules
  storage:
    type : local
    local:
      directory: /tmp/rules
schema:
  configs: []
server:
  grpc_listen_port: 9095
  grpc_server_max_concurrent_streams: 1000
  grpc_server_max_recv_msg_size: 104857600
  grpc_server_max_send_msg_size: 104857600
  http_listen_port: 8080
storage:
  azure:
    account_key: null
    account_name: null
    container_name: null
  cassandra:
    addresses: null
    auth: true
    keyspace: cortex
    password: null
    username: null
  engine: blocks
  index_queries_cache_config:
    memcached:
      expiration: 1h
    memcached_client:
      timeout: 1s
```

Helm chart override values:
```
        sidecar:
          resources:
            limits:
              cpu: 100m
              memory: 100Mi
            requests:
              cpu: 50m
              memory: 50Mi
          rules:
            enabled: true
            searchNamespace: cortex-rules
          alertmanager:
            enabled: true
            searchNamespace: cortex-alertmanager
        useExternalConfig: true
        externalConfigVersion: x
        tags:
          blocks-storage-memcached: true
        config:
          storage:
            engine: blocks
        ingress:
          enabled: true
          annotations:
            kubernetes.io/ingress.class: nginx
          hosts:
            - host: <redacted>
              paths:
                - /
          tls:
            - hosts:
              - <redacted>
        store_gateway:
          replicas: 6
          persistentVolume:
            storageClass: managed-premium
          # store_gateway was crashing on OOMKilled error for large datasets, default was 1Gi limit
          resources:
            limits:
              memory: 7Gi
            requests:
              memory: 4Gi
        compactor:
          persistentVolume:
            size: 256Gi
            storageClass: managed-premium
          # compactor was crash looping on OOMKilled error, default was 1Gi limit
          resources:
            limits:
              memory: 10Gi
            requests:
              memory: 5Gi
        querier:
          resources:
            limits:
              memory: 7Gi
            requests:
              memory: 4Gi
        ingester:
          # WAL needs to be stored to a persistent disk which can survive in the event of an ingester failure
          statefulSet:
            enabled: true
          persistentVolume:
            enabled: true
            size: 64Gi
            storageClass: managed-premium
          # crash looping on OOMKilled error, default was 1Gi limit
          resources:
            limits:
              memory: 12Gi
            requests:
              memory: 12Gi
          extraArgs:
            ingester.max-series-per-metric: 100000
        ruler:
          extraArgs:
            log.level: debug
        alertmanager:
          enabled: true
          statefulSet:
            enabled: true
          persistentVolume:
            size: 8Gi
            storageClass: managed-premium
```

Config map with a rule:
```
apiVersion: v1
data:
  high-cpu.yaml: |-
    groups:
      - name: high-cpu
        rules:
          - alert: HighCPUusage
            expr: avg(100 - rate(node_cpu_seconds_total{mode="idle"}[5m]) * 100) by (instance) > 95
            for: 30m
            labels:
              severity: warning
            annotations:
              description: Metrics from {{ $labels.job }} on {{ $labels.instance }} show CPU > 95% for 30m.
              title: Node {{ $labels.instance }} has high CPU usage
kind: ConfigMap
metadata:
  annotations:
    k8s-sidecar-target-directory: /tmp/rules/fake
  creationTimestamp: "2021-05-23T03:46:06Z"
  labels:
    argocd.argoproj.io/instance: grafana-telemetry-rules
    cortex_rules: "1"
  name: rules-cortex-9f99md47tc
  namespace: cortex-rules
  resourceVersion: "25639302"
  uid: 83fb43df-b3c8-4840-ad54-b11a4a1494f0
```

gburton1 commented on values.yaml:
```diff
@@ -1477,3 +1477,54 @@ configsdb_postgresql:
       key:
+
+## Sidecars that collect the configmaps with specified label and stores the included files them into the respective folders
+sidecar:
```
👍 I like the idea of changing it to `.Values.ruler.sidecar` and `.Values.alertmanager.sidecar`. It only introduces a little bit of duplication, but even that duplication could be useful, e.g. using a different version of sidecar for ruler and alertmanager. We'll change that structure on Monday.

ShuzZzle commented on CHANGELOG.md:
```diff
@@ -6,6 +6,7 @@
 * [BUGFIX] Fixed the default label used in pod affinity expression #162
 * [BUGFIX] Fix label and annotation overrides for services (thanks @kwangil-ha) #164
 
+* [FEATURE] Support dynamic configuration of Ruler and AlertManager using sidecar #150
```
the order of entries should be [CHANGE], [FEATURE], [ENHANCEMENT], [BUGFIX]

amouhadi replied:
updated the order

ShuzZzle commented on templates/alertmanager/alertmanager-dep.yaml:
```diff
@@ -50,6 +50,54 @@ spec:
+          volumeMounts:
+            - name: storage
+              mountPath: "/data"
+              subPath: {{ .Values.alertmanager.persistence.subPath }}
```
hmm that should not be here. If anything it should be i.e.
```YAML
.Values.alertmanager.sidecar.persistence.subPath
```

amouhadi replied:
you're right, we don't need the subpath actually, deleted it

ShuzZzle commented on templates/alertmanager/alertmanager-statefulset.yaml:
```diff
@@ -99,6 +99,61 @@ spec:
+          volumeMounts:
+            - name: storage
+              mountPath: "/data"
+              {{- if .Values.alertmanager.persistentVolume.subPath }}
```
hmm that should not be here. If anything it should be i.e.
```YAML
.Values.alertmanager.sidecar.persistence.subPath
```

amouhadi replied:
you're right, we don't need the subpath actually, deleted it

vinitmasaun commented:
I have tried to incorporate the Alertmanager sidecar for configuration but the sidecar container is throwing the following error I have deployed the ConfigMap to the same namespace. This is on AWS EKS version 1.19:


[2022-01-04 23:23:00] MaxRetryError when calling kubernetes: HTTPConnectionPool(host='localhost', port=80): Max retries exceeded with url: /api/v1/namespaces/cortex/configmaps?labelSelector=app%3Dcortex-alertmanager&watch=True (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f7f40ed3040>: Failed to establish a new connection: [Errno 111] Connection refused'))

Following is my ConfigMap yaml:

apiVersion: v1
kind: ConfigMap
metadata:
  name: cortex-alertmanager
  namespace: cortex
  labels:
    app: cortex-alertmanager
data:
  alertmanager_config: |
    route:
      group_by:
        - alertname
        - instance
        - severity
        - job
        - namespace
      group_wait: 30s
      group_interval: 5m
      repeat_interval: 1h
      receiver: 'splunk_webhook'
      routes:
        - receiver: 'splunk_webhook'
          continue: true
    receivers:
      - name: 'splunk_webhook'
        webhook_configs:
          - url: 'redacted'
            send_resolved: true
            http_config:
              basic_auth:
                username: 'redacted'
                password: 'redacted'

    inhibit_rules:
      - source_match:
          severity: 'critical'
        target_match:
          severity: 'warning'
        equal: ['alertname']

gburton1 commented:
`localhost:80` is not going to be the right place to reach the kubernetes API. It should be going through the `kubernetes` service in the `default` namespace. And that should resolve to something like 10.0.0.1, so your requests from the sidecar should be something like: `GET https://10.0.0.1:443/api/v1...`

We have been running the AlertManager and Ruler sidecar configurations in production for a while now.

vinitmasaun commented:
> `localhost:80` is not going to be the right place to reach the kubernetes API. It should be going through the `kubernetes` service in the `default` namespace. And that should resolve to something like 10.0.0.1, so your requests from the sidecar should be something like: `GET https://10.0.0.1:443/api/v1...`
> 
> We have been running the AlertManager and Ruler sidecar configurations in production for a while now.

Right, but how can I control the kube API url that the sidecar is constructing?  Is there a values.yaml entry I should be setting for this in addition to the sidecar section?  I am using cortex helm chart version 1.2.0.  Following is my values.yaml snippet for Alertmanager sidecar:

alertmanager:
  replicas: 1
  statefulSet:
    enabled: true
  sidecar:
    image:
      repository: quay.io/kiwigrid/k8s-sidecar
    enabled: true
    searchNamespace: cortex
    skipTlsVerify: true
    label: Name
    labelValue: cortex-alertmanager
    resources:
      limits:
        cpu: 100m
        memory: 100Mi
      requests:
        cpu: 50m
        memory: 50Mi
  persistentVolume:
    enabled: true

vinitmasaun commented:
> 

it looks like the issue with the localhost:80 is with a bug in the kiwigrid/k8s-sidecar image tag 1.10.1 (https://github.com/kiwigrid/k8s-sidecar/issues/114) I updated the tag value in my values.yaml to 1.11.1 and that resolved the kube api url issue however: now I am getting a permission denied error when the ConfigMap is downloaded and the sidecar is trying to store it in the data folder:

[2022-01-05 02:04:35] Working on ADDED configmap cortex/cortex-alertmanager
[2022-01-05 02:04:35] Received unknown exception: [Errno 13] Permission denied: '/data/alertmanager_config'


I've tried changing the following values in the values.yaml but neither seem to resolve the permission denied error:

alertmanager:
  replicas: 1
  statefulSet:
    enabled: true
  sidecar:
    image:
      repository: quay.io/kiwigrid/k8s-sidecar
      tag: 1.11.1
    enabled: true
    searchNamespace: cortex
    skipTlsVerify: true
    label: cortex_alertmanager
    resources:
      limits:
        cpu: 100m
        memory: 100Mi
      requests:
        cpu: 50m
        memory: 50Mi
    containerSecurityContext:
      **enabled: false
      readOnlyRootFilesystem: false**
  persistentVolume:
    enabled: true
